programs/server/Server.cpp

```diff
@@ -1968,25 +1969,16 @@ void Server::updateServers(
                 LOG_INFO(log, "Stopped listening for {}", server.getDescription());
             }
         }
+    }
 
     createServers(config, listen_hosts, listen_try, server_pool, async_metrics, servers, /* start_servers: */ true);
 
-    /// Remove servers once all their connections are closed
-    while (std::any_of(servers.begin(), servers.end(), [](const auto & server) { return server.isStopping(); }))
+    /// Remove servers without waiting for active connections,
+    /// since they will be served anyway via fd that returned by accept().
+    std::erase_if(servers, [](auto & server)
     {
-        std::this_thread::sleep_for(std::chrono::milliseconds(100));
-        std::erase_if(servers, [&log](auto & server)
-        {
-            if (!server.isStopping())
-                return false;
-            auto is_finished = server.currentConnections() == 0;
-            if (is_finished)
-                LOG_DEBUG(log, "Server finished: {}", server.getDescription());
-            else
-                LOG_TRACE(log, "Waiting server to finish: {}", server.getDescription());
-            return is_finished;
-        });
-    }
+        return server.isStopping();
+    });
```

Comment:
It looks unsafe to delete servers without letting them finish their work.

Reply:
Maybe we could just keep servers in this vector which are stopping but haven't stopped yet

Reply:
Yes, it is true right now, but I don't think that it is a good thing, since TCPHandler should not break the connection even if server had been changed, there is simply no reason for this since it operates on a separate fd, that had been allocated via accept().

But right now it requires server to check that it hadn't been stopped, while I think that checking `BaseDaemon::isCancelled()` is enough, and better.

Reply:
So I've rewrite server handlers to avoid depends on server instances, so now it should be safe to remove servers with active handlers, for more details see the commit descriptions (in particular the last patch in the PR).